### PR TITLE
feat(ubuntu):Add an `amazon-ebs` AMI builder

### PR DESCRIPTION
**Pull request by jayfranco999**
As per https://github.com/jenkins-infra/helpdesk/issues/4353#issue-2586531372

Related to https://github.com/jenkins-infra/helpdesk/issues/4316#issuecomment-2410363812

This PR introduces the following to packer-images:

- Add amazon-ebs builder logic in Jenkinsfile_k8s
- Include EC2 source for build-jenkins-agent-ubuntu.pkr.hcl
 - Add datasources.pkr.hcl to look-up ubuntu-22_04 AMI id for respective architecture
 - Define aws_instance_types in locals.pkr.hcl
 - Include Amazon plugin for packer-images
 - Define all the common settings for AWS base image AMI's in sources.pkr.hcl
 - Add AWS related packer-image variables (AWS region and credentials)

### CODE REVIEW

File: variables.pkr.hcl [hunk 1]
```diff
@@ -13,6 +13,18 @@ variable "architecture" {
...
+  default = "us-east-2"
+}
+variable "aws_access_key_id" {
+  type    = string
+  default = env("AWS_ACCESS_KEY_ID")
+}
+variable "aws_secret_access_key" {
+  type    = string
+  default = env("AWS_SECRET_ACCESS_KEY")
+}
```
Comment (dduportal) [inline variables.pkr.hcl:27]:
```suggestion
```

as discussed together, we have to use https://developer.hashicorp.com/packer/integrations/hashicorp/amazon#environment-variables

File: sources.pkr.hcl [hunk 2]
```diff
@@ -1,3 +1,47 @@
+# This source defines all the common settings for any AWS AMI (whatever Operating System)
+source "amazon-ebs" "base" {
+  # profile       = "terraform-developer"
+
+  # AWS API connection
+  access_key      = var.aws_access_key_id
+  secret_key      = var.aws_secret_access_key
+
```
Comment (dduportal) [inline sources.pkr.hcl:8]:
```suggestion
```

as discussed together, we have to use https://developer.hashicorp.com/packer/integrations/hashicorp/amazon#environment-variables

File: datasources.pkr.hcl [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  secret_key      = var.aws_secret_access_key
+  filters = {
+    name                = "Windows_Server-2022-English-Core-Base-*"
+    root-device-type    = "ebs"
+    virtualization-type = "hvm"
+  }
+  most_recent = true
+  owners      = ["amazon"]
+  region      = var.aws_region
+}
```
Comment (dduportal) [inline datasources.pkr.hcl:41]:
```suggestion
```

Not needed in this PR: we don't want to clobber setups with unused codes (neither with comments)

File: datasources.pkr.hcl [hunk 4]
```diff
@@ -0,0 +1,41 @@
+# # Data sources are always treated BEFORE locals and sources.
+data "amazon-ami" "ubuntu-22_04" {
+  access_key      = var.aws_access_key_id
+  secret_key      = var.aws_secret_access_key
```
Comment (dduportal) [inline datasources.pkr.hcl:4]:
```suggestion
```

as discussed together, we have to use https://developer.hashicorp.com/packer/integrations/hashicorp/amazon#environment-variables

**Review by dduportal — DISMISSED** [review]
- Updatecli check fails: as the manifest assumes the code is on the principal branch, either you add it on a subsequent PR or we can accept it's working once merged (but you'll have to test it locally without the `scmid: ` keys)
- We must use environment variable for the AWS provider setup. Please read carefully https://developer.hashicorp.com/packer/integrations/hashicorp/amazon#environment-variables and https://developer.hashicorp.com/packer/integrations/hashicorp/amazon#static-credentials and mùake sure you get the difference (see suggestions): it's a blocker

**Comment by jayfranco999** [comment]
Update: 

`ubuntu-22_04` AMI images were successfully built with packer-images – https://infra.ci.jenkins.io/job/infra-tools/job/packer-images/job/PR-1447/5/ 

Since windows AMI image source logic is not included in this PR, the windows agent builds are failing in `PACKER_LOG=1 packer validate ./` stage of the build pipeline with exit-code-1

```
[2024-10-15T12:27:00.527Z]   on sources.pkr.hcl line 33:
[2024-10-15T12:27:00.527Z]   (source code not available)
[2024-10-15T12:27:00.527Z] 
[2024-10-15T12:27:00.527Z] with data.amazon-ami as map of object with 1 element,
[2024-10-15T12:27:00.527Z]      local.agent_os_version_safe as "2022",
[2024-10-15T12:27:00.527Z]      var.agent_os_type as "windows".
[2024-10-15T12:27:00.527Z] 
[2024-10-15T12:27:00.527Z] The given key does not identify an element in this collection value.
[2024-10-15T12:27:00.527Z] 
[2024-10-15T12:27:00.527Z] Error: Invalid index
[2024-10-15T12:27:00.527Z] 
[2024-10-15T12:27:00.527Z]   on sources.pkr.hcl line 33:
[2024-10-15T12:27:00.527Z]   (source code not available)
[2024-10-15T12:27:00.527Z] 
[2024-10-15T12:27:00.527Z] with data.amazon-ami as map of object with 1 element,
[2024-10-15T12:27:00.527Z]      local.agent_os_version_safe as "2022",
[2024-10-15T12:27:00.527Z]      var.agent_os_type as "windows".
[2024-10-15T12:27:00.527Z] 
[2024-10-15T12:27:00.527Z] The given key does not identify an element in this collection value.
```
I believe the problem is with `source_ami = data.amazon-ami["${var.agent_os_type}-${local.agent_os_version_safe}"].id` command in `sources.pkr.hcl` that causes errors in the packer validate stage

File: Jenkinsfile_k8s [hunk 5]
```diff
@@ -149,6 +175,9 @@ pipeline {
           }
         }
         environment {
+          // Defines the following environment variables: AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY
+          AWS_ACCESS_KEY_ID             = credentials('packer-aws-access-key-id')
+          AWS_SECRET_ACCESS_KEY         = credentials('packer-aws-secret-access-key')
```
Comment (dduportal) [inline Jenkinsfile_k8s:180]:
```suggestion
          // Defines the following environment variables: AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY
          AWS_ACCESS_KEY_ID             = credentials('packer-aws-access-key-id')
          AWS_SECRET_ACCESS_KEY         = credentials('packer-aws-secret-access-key')
          AWS_DEFAULT_REGION = 'us-east-2'
```

File: build-jenkins-agent-ubuntu.pkr.hcl [hunk 6]
```diff
@@ -3,6 +3,14 @@ build {
     name = "ubuntu"
   }
 
+  source "amazon-ebs.base" {
+    name         = "ubuntu"
+    ssh_username = "ubuntu"
+    # Egg-and-chicken: what is the base image to start from (eg. what is my egg)?
+    source_ami = local.aws_ubuntu_amis[var.architecture]
+    region = env("AWS_DEFAULT_REGION")
```
[the remark this answers is not in the document]
Comment (dduportal) [reply build-jenkins-agent-ubuntu.pkr.hcl:11]: See https://github.com/jenkins-infra/packer-images/pull/1447/files#r1803089067 suggestion

File: locals.pkr.hcl [hunk 7]
```diff
@@ -4,6 +4,21 @@ locals {
   agent_os_version_safe = replace(var.agent_os_version, ".", "_")
   image_name            = format("jenkins-agent-%s-%s-%s", var.agent_os_type, var.agent_os_version, var.architecture)
   unique_image_name     = format("%s-%s", local.image_name, local.now_unix_timestamp)
+  
+  # aws_spot_instance_types = { // for spot instances
+  #   # 4 vCPU x86 / 16 GB / $0.1670 - https://aws.amazon.com/en/ec2/instance-types/t3/#Product_Details
+  #   "amd64" = ["t3.xlarge", "t3a.xlarge", "t2.xlarge", "m6a.xlarge"]
+  #   # 4 vCPU ARM64 (Gravitnb)/ 16 GB / $0.1344 - https://aws.amazon.com/en/ec2/instance-types/t4/#Product_Details
+  #   "arm64" = ["t4g.xlarge", "m7g.xlarge"]
+  # }
```
Comment (dduportal) [inline locals.pkr.hcl:13]:
```suggestion
```

Don't forget to remove these comments

File: locals.pkr.hcl [hunk 8]
```diff
@@ -4,6 +4,21 @@ locals {
...
+  aws_instance_types = {
+    "amd64" = "t3.xlarge"
+    "arm64" = "t4g.xlarge"
+  }
+  aws_ubuntu_amis = {
+    "amd64" = "ami-00eb69d236edcfaf8"
+    "arm64" = "ami-039e419d24a37cb82"
```
Comment (dduportal) [inline locals.pkr.hcl:20]: TODO: add a comment with the `aws` CLI used to retrieve the AMI ID for both cases

File: main.pkr.hcl [hunk 9]
```diff
@@ -1,6 +1,10 @@
 packer {
   required_version = ">= 1.7.2, < 2"
   required_plugins {
+    amazon = {
+      version = "1.3.3"
```
Comment (dduportal) [inline main.pkr.hcl:5]:
```suggestion
      # TODO: track this version with `updatecli`
      version = "1.3.3"
```

File: locals.pkr.hcl [hunk 10]
```diff
@@ -4,6 +4,21 @@ locals {
...
+  #   "amd64" = ["t3.xlarge", "t3a.xlarge", "t2.xlarge", "m6a.xlarge"]
+  #   # 4 vCPU ARM64 (Gravitnb)/ 16 GB / $0.1344 - https://aws.amazon.com/en/ec2/instance-types/t4/#Product_Details
+  #   "arm64" = ["t4g.xlarge", "m7g.xlarge"]
+  # }
+  aws_instance_types = {
+    "amd64" = "t3.xlarge"
+    "arm64" = "t4g.xlarge"
+  }
+  aws_ubuntu_amis = {
+    "amd64" = "ami-00eb69d236edcfaf8"
```
Comment (dduportal) [inline locals.pkr.hcl:19]:
```suggestion
          # TODO: track this version with `updatecli`
    "amd64" = "ami-00eb69d236edcfaf8"
```

File: locals.pkr.hcl (hunk 8 above)
Comment (dduportal) [inline locals.pkr.hcl:20]:
```suggestion
    # TODO: track this version with `updatecli`
    "arm64" = "ami-039e419d24a37cb82"
```

File: sources.pkr.hcl [hunk 11]
```diff
@@ -1,3 +1,40 @@
+# This source defines all the common settings for any AWS AMI (whatever Operating System)
+source "amazon-ebs" "base" {
+  # profile       = "terraform-developer"
+
+  ami_name      = "${local.image_name}-${var.architecture}-${local.now_unix_timestamp}"
+  # spot_instance_types = local.aws_spot_instance_types[var.architecture] // if spot instances is used
+  # spot_price          = "auto"
+  instance_type = local.aws_instance_types[var.architecture]
+  # instance_type = "t2.xlarge"
```
Comment (dduportal) [inline sources.pkr.hcl:9]:
```suggestion
  instance_type = local.aws_instance_types[var.architecture]
```

File: sources.pkr.hcl [hunk 12]
```diff
@@ -1,3 +1,40 @@
...
+
+
+  # Define custom rootfs for build to avoid later filesystem extension during agent startups
+  launch_block_device_mappings {
+    delete_on_termination = true
+    device_name           = "/dev/sda1"
+    volume_size           = local.windows_disk_size_gb
+    volume_type           = "gp2"
```
Comment (dduportal) [inline sources.pkr.hcl:17]:
```suggestion
    # TODO: check if we can use `gp3` (blocker was ec2 plugin, not packer)
    volume_type           = "gp2"
```

File: variables.pkr.hcl [hunk 13]
```diff
@@ -13,6 +13,10 @@ variable "architecture" {
   description = "CPU architecture ID of the build with the following possible values: [amd64 (default), arm64]"
   default     = "amd64"
 }
+variable "aws_destination_region" {
+  type    = string
+  default = env("AWS_DEFAULT_REGION")
```
Comment (dduportal) [inline variables.pkr.hcl:18]:
```suggestion
  # Defaults to the only region we use
  default = env("AWS_DEFAULT_REGION")
```

**Comment by dduportal** [comment]
@jayfranco999 Proposal for a clear title:

```
feat(ubuntu):Add an `amazon-ebs` AMI builder
```

instead of:

```
feat!(aws):Add Amazon-ebs builder for Linux AMIs
```

- It is a new feature, not a breaking change at all
- AWS is not a component of the repository. Ubuntu is.

WDYT?

**Review by dduportal — APPROVED** [review]

**Pull request merged**